flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,380 @@
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) throws IOException {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        activeChangeSetSize += value.length;
+        if (activeChangeSetSize >= preEmptivePersistThresholdInBytes) {
+            LOG.debug(
+                    "pre-emptively flush {}Mb of appended changes to the common store",
+                    activeChangeSetSize / 1024 / 1024);
+            persistInternal(notUploaded.isEmpty() ? activeSequenceNumber : notUploaded.firstKey());
+        }
+    }
+
+    @Override
+    public SequenceNumber initialSequenceNumber() {
+        return INITIAL_SQN;
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        LOG.trace("query {} sqn: {}", logId, activeSequenceNumber);
+        SequenceNumber tmp = activeSequenceNumber;
+        rollover();
+        return tmp;
+    }
+
+    @Override
+    public CompletableFuture<ChangelogStateHandleStreamImpl> persist(SequenceNumber from)
+            throws IOException {
+        LOG.debug(
+                "persist {} starting from sqn {} (incl.), active sqn: {}",
+                logId,
+                from,
+                activeSequenceNumber);
+        return persistInternal(from);
+    }
+
+    private CompletableFuture<ChangelogStateHandleStreamImpl> persistInternal(SequenceNumber from)
+            throws IOException {
+        synchronized (lock) {
+            ensureCanPersist(from);
+            rollover();
+            Map<SequenceNumber, StateChangeSet> tailMap = notUploaded.tailMap(from, true);
+            Map<SequenceNumber, StateChangeSet> toUpload = new HashMap<>(tailMap);
+            tailMap.clear(); // prevent re-uploads
+            NavigableMap<SequenceNumber, UploadResult> readyToReturn = uploaded.tailMap(from, true);
+            LOG.debug("collected readyToReturn: {}, toUpload: {}", readyToReturn, toUpload);
+
+            SequenceNumberRange range = SequenceNumberRange.generic(from, activeSequenceNumber);
+            if (range.size() == readyToReturn.size()) {
+                checkState(toUpload.isEmpty());
+                return completedFuture(buildHandle(keyGroupRange, readyToReturn));
+            } else {
+                CompletableFuture<ChangelogStateHandleStreamImpl> future =
+                        new CompletableFuture<>();
+                uploadCompletionListeners.add(
+                        new UploadCompletionListener(keyGroupRange, range, readyToReturn, future));
+                if (!toUpload.isEmpty()) {
+                    uploader.upload(
+                            new UploadTask(
+                                    toUpload.values(),
+                                    this::handleUploadSuccess,
+                                    this::handleUploadFailure));
```

Comment:
Do we need to make this class semi thread safe for handling the uploading results? Can not they go through mailbox for example? As far as I can tell only those two callbacks (`handleUploadSuccess` and `handleUploadFailure`) are called from different threads?

Reply:
I think it's doable and I left a comment about it:
```
    /** Lock to synchronize handling of upload completion with new upload requests. */
    // todo: replace with mailbox executor (after FLINK-23204)
    private final Object lock = new Object();
```
But as FLINK-23204 is currently not implemented (and there is no consensus on how to do that), I'd like not to add such a dependency.